File: swim-system-java/swim-core-java/swim.structure/src/main/java/swim/structure/InterpreterSettings.java
// Copyright 2015-2021 Swim inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.structure;

import swim.codec.Debug;
import swim.codec.Format;
import swim.codec.Output;
import swim.util.Murmur3;

public class InterpreterSettings implements Debug {

  protected final int maxScopeDepth;

  public InterpreterSettings(int maxScopeDepth) {
    this.maxScopeDepth = maxScopeDepth;
  }

  public final int maxScopeDepth() {
    return this.maxScopeDepth;
  }

  public InterpreterSettings maxScopeDepth(int maxScopeDepth) {
    return this.copy(maxScopeDepth);
  }

  protected InterpreterSettings copy(int maxScopeDepth) {
    return new InterpreterSettings(maxScopeDepth);
  }

  protected boolean canEqual(Object other) {
    return other instanceof InterpreterSettings;
  }

  @Override
  public boolean equals(Object other) {
    if (this == other) {
      return true;
    } else if (other instanceof InterpreterSettings) {
      final InterpreterSettings that = (InterpreterSettings) other;
      return that.canEqual(this) && this.maxScopeDepth == that.maxScopeDepth;
    }
    return false;
  }

  private static int hashSeed;

  @Override
  public int hashCode() {
    if (InterpreterSettings.hashSeed == 0) {
      InterpreterSettings.hashSeed = Murmur3.seed(InterpreterSettings.class);
    }
    return Murmur3.mash(Murmur3.mix(InterpreterSettings.hashSeed, this.maxScopeDepth));
  }

  @Override
  public <T> Output<T> debug(Output<T> output) {
    output = output.write("new").write(' ').write("InterpreterSettings").write('(')
                   .debug(this.maxScopeDepth).write(')');
    return output;
  }

  @Override
  public String toString() {
    return Format.debug(this);
  }

  public static final int MAX_SCOPE_DEPTH;

  private static InterpreterSettings standard;

  public static InterpreterSettings standard() {
    if (InterpreterSettings.standard == null) {
      InterpreterSettings.standard = new InterpreterSettings(InterpreterSettings.MAX_SCOPE_DEPTH);
    }
    return InterpreterSettings.standard;
  }

  static {
    int maxScopeDepth;
    try {
      maxScopeDepth = Integer.parseInt(System.getProperty("swim.structure.interpreter.max.scope.depth"));
    } catch (NumberFormatException e) {
      maxScopeDepth = 1024;
    }
    MAX_SCOPE_DEPTH = maxScopeDepth;
  }

}

File: swim-system-js/swim-core-js/@swim/structure/main/InterpreterSettings.ts
// Copyright 2015-2020 Swim inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {HashCode, Murmur3} from "@swim/util";
import {Output, Format, Debug} from "@swim/codec";

/**
 * Either an [[InterpreterSettings]] instance, or an [[InterpreterSettingsInit]]
 * object initializer.
 */
export type AnyInterpreterSettings = InterpreterSettings | InterpreterSettingsInit;

/**
 * [[InterpreterSettings]] object initializer.
 */
export interface InterpreterSettingsInit {
  lineSeparator?: string | null;
  isPretty?: boolean;
  isStyled?: boolean;
}

export class InterpreterSettings implements Debug, HashCode {
  /** @hidden */
  readonly _maxScopeDepth: number;

  constructor(maxScopeDepth: number) {
    this._maxScopeDepth = maxScopeDepth;
  }

  maxScopeDepth(): number;

  maxScopeDepth(maxScopeDepth: number): InterpreterSettings;

  maxScopeDepth(maxScopeDepth?: number): number | InterpreterSettings {
    if (maxScopeDepth === void 0) {
      return this._maxScopeDepth;
    } else {
      return this.copy(maxScopeDepth);
    }
  }

  protected copy(maxScopeDepth: number): InterpreterSettings {
    return new InterpreterSettings(maxScopeDepth);
  }

  protected canEqual(that: unknown): boolean {
    return that instanceof InterpreterSettings;
  }

  equals(that: unknown): boolean {
    if (this === that) {
      return true;
    } else if (that instanceof InterpreterSettings) {
      return that.canEqual(this) && this._maxScopeDepth === that._maxScopeDepth;
    }
    return false;
  }

  hashCode(): number {
    if (InterpreterSettings._hashSeed === void 0) {
      InterpreterSettings._hashSeed = Murmur3.seed(InterpreterSettings);
    }
    return Murmur3.mash(Murmur3.mix(InterpreterSettings._hashSeed, this._maxScopeDepth));
  }

  debug(output: Output): void {
    output = output.write("new").write(32/*' '*/).write("InterpreterSettings")
        .write(40/*'('*/).debug(this._maxScopeDepth).write(41/*')'*/);
  }

  toString(): string {
    return Format.debug(this);
  }

  private static _hashSeed?: number;
  private static _standard?: InterpreterSettings;

  static standard(): InterpreterSettings {
    if (InterpreterSettings._standard === void 0) {
      const maxScopeDepth = 1024;
      InterpreterSettings._standard = new InterpreterSettings(maxScopeDepth);
    }
    return InterpreterSettings._standard;
  }
}
